### PR TITLE
Allowing larger segment size for SOH

### PR DESCRIPTION
**Pull request by Maoni0**
in preparation for enabling it for NoGC region. the main change here is to make sure we do an ephemeral promote if the new ephemeral size cannot fit into a normal sized seg.

### CODE REVIEW

**Comment by Maoni0** [comment]
@adityamandaleeka PTAL

File: src/gc/gc.cpp [hunk 1]
```diff
@@ -2963,8 +2966,12 @@ gc_heap::dt_low_ephemeral_space_p (gc_tuning_point tp)
             
             dprintf (GTC_LOG, ("h%d: plan eph size is %Id, new gen0 is %Id", 
                 heap_number, plan_ephemeral_size, new_gen0size));
-            ret = ((size_t)(heap_segment_reserved (ephemeral_heap_segment) - (heap_segment_mem (ephemeral_heap_segment))) <
-                    (plan_ephemeral_size + new_gen0size));
+
+            // While we were in the no gc region we could have allocated a larger than normal segment,
```
Comment (adityamandaleeka) [inline src/gc/gc.cpp]: > While we were in the no gc region [...]

This sounds like we are assuming that we know we got here after being in a no-GC region. Is that always the case?

I just did a quick search and it looks like we only call `dt_low_ephemeral_space_p` with `tuning_deciding_promote_ephemeral` if the pause mode is _not_ `pause_no_gc` (in `gc_heap::expand_heap`). I think I'm missing something here.

File: src/gc/gc.cpp [hunk 2]
```diff
@@ -9928,6 +9926,15 @@ HRESULT gc_heap::initialize_gc (size_t segment_size,
 #endif //BACKGROUND_GC
 #endif //WRITE_WATCH
 
+#ifdef BACKGROUND_GC
+    //leave the first page to contain only segment info
```
Comment (adityamandaleeka) [inline src/gc/gc.cpp]: minor nit: comment formatting isn't consistent.

File: src/gc/gcpriv.h [hunk 3]
```diff
@@ -3002,6 +3002,17 @@ class gc_heap
     PER_HEAP_ISOLATED
     size_t min_segment_size;
 
+    // For SOH we always allocate segments of the same
+    // size unless no gc region requires larger ones.
```
Comment (adityamandaleeka) [inline src/gc/gcpriv.h]: Please change 'no gc region' to 'NoGCRegion'. It will be way clearer when parsing this comment later :smile:.

File: src/gc/gc.cpp (hunk 1 above)
Comment (Maoni0) [inline src/gc/gc.cpp]: We got here after we've *exited* from the no gc region mode, now we need to handle the seg we allocated "While we were in the no gc region".

File: src/gc/gcpriv.h (hunk 3 above)
Comment (Maoni0) [inline src/gc/gcpriv.h]: the actual implementation is all named something no_gc_region, that's why I referred to it as no gc region. I guess I can change it to no_gc_region if that seems better.

Comment (adityamandaleeka) [inline src/gc/gcpriv.h]: Yes, that works too. The way it is now, it sounds like something totally different.

**Comment by adityamandaleeka** [comment]
The a comment on `get_valid_segment_size` talks about having places where that function is called unnecessarily. Since you've removed all (?) of those places, we can probably remove the part of the comment now, right? It also says "TODO: move back after we do this"--not sure what that means.

File: src/gc/gc.cpp (hunk 1 above)
Comment (adityamandaleeka) [inline src/gc/gc.cpp]: But we also end up here if there was a heap expansion without ever having encountered a no-gc region, right?

Comment (Maoni0) [inline src/gc/gc.cpp]: yeah, but that doesn't cause us to ever create a segment that's larger than the default.

**Comment by Maoni0** [comment]
I will remove the comments for get_valid_segment_size.

File: src/gc/gcpriv.h (hunk 3 above)
Comment (Maoni0) [inline src/gc/gcpriv.h]: I am curious - now that you know the  implementation is called no_gc_region, does this still sound like something totally different?

**Comment by Maoni0** [comment]
@dotnet-bot test Ubuntu x64 Formatting

**Comment by Maoni0** [comment]
@dotnet-bot test OSX10.12 x64 Checked Build and Test

File: src/gc/gcpriv.h (hunk 3 above)
Comment (adityamandaleeka) [inline src/gc/gcpriv.h]: Yes. I knew what this comment was trying to say (we are, after all, in a PR about making changes in preparation for no-gc region), but my point is that it's ambiguous and needlessly hard to parse. Your suggestion about adding the underscores is fine.

File: src/gc/gc.cpp (hunk 1 above)
Comment (adityamandaleeka) [inline src/gc/gc.cpp]: OK. As with the other comment, I think this needs to be rephrased in order to make it clear outside of the context of this PR. Can you rephrase this to something like "in cases where we were in a no-gc region..."?

"While we were in the no gc region" implies that we had to have been in a no-gc region, which is not necessarily the case.

**Comment by Maoni0** [comment]
updated commit to address all comment feedback.

**Comment by adityamandaleeka** [comment]
LGTM, thanks.
